Automation/Congestion fee Implementation

nizam-supraoracles opened the pull request:
Implementation of https://github.com/Entropy-Foundation/smr-moonshot/issues/1421

nizam-supraoracles committed:
automation automation - charge `flat_registration_fee` at the time of registration which is non refundable

nizam-supraoracles committed:
automation registry - on_new_epoch fee charges for an task

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -255,17 +272,86 @@ module supra_framework::automation_registry {
+        let task_max_gas = (task.max_gas_amount as u256);
+
+        let epoch_interval = aei.epoch_interval;
+        let remaining_time = task.expiry_time - current_time;
```
Is this guaranteed to be subtraction safe at this point? @nizam-supraoracles

aregng replied:
yes, as the expired tass have been already removed from registry before charging for the new epoch for the remain tasks.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -255,17 +272,86 @@ module supra_framework::automation_registry {
+
+        let epoch_interval = aei.epoch_interval;
+        let remaining_time = task.expiry_time - current_time;
+        let min_interval = if (remaining_time < epoch_interval) remaining_time else epoch_interval;
```
Use `min` from library

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -255,17 +272,86 @@ module supra_framework::automation_registry {
+        // Compute the base automation fee (taf)
+        let gas_proportion = (task_max_gas * DECIMAL / max_gas_cap);
+        let base_fee = (abf * gas_proportion) / DECIMAL;
+        let taf = base_fee * (min_interval as u256); // Total base fee for the interval
```
We can simplify this as follows
`taf = abf * task_max_gas * min_interval / max_gas_cap`

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -255,17 +272,86 @@ module supra_framework::automation_registry {
+
+        // Compute the congestion fee per task (tcf)
+        let tcf = if (acf > 0) {
+            (acf * gas_proportion * (min_interval as u256)) / DECIMAL
```
Here also
tcf = if (acf>0) { acf * task_max_gas * min_interval}

aregng replied:
do we need `if (acf > 0)` here ? Is it for the optimization not to do multiplication at all ?

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -255,17 +272,86 @@ module supra_framework::automation_registry {
+        } else 0;
+
+        let transfer_fee_amount = (taf + tcf as u64);
+        supra_account::transfer(&create_signer(task.owner), registry_fee_address, transfer_fee_amount)
```
@nizam-supraoracles, @aregng, We MUST NOT use any operation which can fail, since this function would be called during epoch change. Check if balance is sufficient, and if not, remove the task from the registry.

aregng replied:
Yes, we have discussed this yesterday.

nizam-supraoracles replied:
If a user doesn't have enough balance and we are removing them from the registry, do we need to recalculate the epoch congestion fee based on that?

sjoshisupra replied:
@nizam-supraoracles , no, we don't recalculate.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -240,6 +248,15 @@ module supra_framework::automation_registry {
         let current_time = timestamp::now_seconds();
         let gas_committed_for_next_epoch = 0;
 
+        // Apply the latest configuration if any parameter has been updated.
```
@nizam-supraoracles, I do not see any logic to refund the amount if the epoch was shorter than expected. Is it somewhere else?

nizam-supraoracles replied:
I am current doing the same

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -255,17 +272,86 @@ module supra_framework::automation_registry {
+            automation_registry,
+            automation_epoch_info,
+            &automation_registry_config,
+            ids,
```
shouldn't this be only ids of active tasks ?

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -255,17 +272,86 @@ module supra_framework::automation_registry {
+        else threshold_usage - threshold_percentage;
+
+        // Compute the automation congestion fee (acf) for the epoch
+        let acf = if (threshold_surplus > 0) {
```
do we need this if, the `acf` will be 0 if `threshold_surplus` is 0, right?

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -255,17 +272,86 @@ module supra_framework::automation_registry {
+        current_time: u64,
+        tcmg: u256
+    ) {
+        let max_gas_cap = (arc.registry_max_gas_cap as u256);
```
Let's have `acf` calculation for task as separate function

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -255,17 +272,86 @@ module supra_framework::automation_registry {
+        ar: &AutomationRegistry,
+        aei: &AutomationEpochInfo,
+        arc: &AutomationRegistryConfig,
+        ids: vector<u64>,
```
I think we can avoid passing the `ids` here which I expect to be ids of all active tasks, if we meet the precondition that `fee-charges` are called only after cleanup has been done and registry contains only active tasks.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -255,17 +272,86 @@ module supra_framework::automation_registry {
     }
 
+    /// Charges automation task fees for all active tasks at the beginning of a new epoch.
+    fun fee_charges_on_new_epoch(
```
I think we need to update this function not to charge the fee but to return 
`{task-id, sender, fee}` tripple information for each task which is ordered by task-id.
Then do a charging gradually checking account balance, if at some point account does not have enough balance to pay automation fee then that task with corresponding id is scheduled to be cancelled.
Please also remember that we need to post 2 type of events:
- one for the charge with above-mentioned tripple information
- one for the potential cancellation due to insufficient balance again with the same  triple information.

nizam-supraoracles replied:
Okay, I have made the changes accordingly. However, two more fields, `max_gas_fee` and `expiry_time`, are required to calculate `gas_committed_for_next_epoch`, so I have added them as well.

nizam-supraoracles committed:
auomation fee charges - create `calculate_automation_congestion_fee` seprate function and address pr comments

nizam-supraoracles committed:
automation registry fee - create new function to handle `process_automation_task_fees` and emit event based on the execusion + few function name update

nizam-supraoracles committed:
automation registry - check if `automation_fee_cap_for_epoch` execeed then cancal this task

nizam-supraoracles committed:
automation registry - make consistance for name of task_index for entire module

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -460,7 +631,7 @@ module supra_framework::automation_registry {
 
-        event::emit(CancelledAutomationTask { id: automation_task_metadata.id });
+        event::emit(CancelledAutomationTask { task_index: automation_task_metadata.task_index });
         refund_automation_task_fee(
```
No refund should be done when cancelling the task.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -175,7 +174,36 @@ module supra_framework::automation_registry {
+        task_index: u64,
+        owner: address,
+        fee: u64,
+        reason: String,
```
Isn't the name of the event itself sufficient? Why do we need `reason:String` ? Is there something that won't be conveyed with the name and other fields?

nizam-supraoracles replied:
Okay, As per latest event names I have removed the reason field

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -237,35 +273,174 @@ module supra_framework::automation_registry {
+    /// Charges automation task fees for a single task at the beginning of a new epoch.
+    /// It's return boolean value
+    ///     true means balance fee charges successfully deducted,
+    ///     false means due to task expiry or
```
Please add a CAUTION comment that this is supposed to be called only after removing expired task and must not be called for expired task since otherwise subtraction safety does not hold.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -237,35 +273,174 @@ module supra_framework::automation_registry {
+                max_gas_amount: task.max_gas_amount,
+                expiry_time: task.expiry_time,
+                automation_fee_cap_for_epoch: task.automation_fee_cap_for_epoch,
+            }
```
This seems inefficient. We calculate `task_fee` here but construct a temporary heavy duty object `AutomationTaskFee`. All the other field values are already available in the registry so why do that? We only need to pass `(task_index,task_fee)` and then in `process_automation_task_fees`, we fetch the required values from the registry. Here, we are creating the object with unnecessary fields. The gas cost for look up in the registry, whether being done here or `process_automation_task_fees` is the same, but we add to the construction cost and then to the object copy cost.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -237,35 +273,174 @@ module supra_framework::automation_registry {
-            if (task.state != CANCELLED && task.expiry_time > (current_time + automation_epoch_info.epoch_interval)) {
-                gas_committed_for_next_epoch = gas_committed_for_next_epoch + task.max_gas_amount;
-            };
+        vector::for_each(ids, |task_index| {
```
let's have it as a function `cleanup_and_activate_tasks(tasks, current_time) -> u64` which return total committed max-gas`

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -237,35 +273,174 @@ module supra_framework::automation_registry {
+                enumerable_map::remove_value(&mut automation_registry.tasks, task_index);
             } else {
                 task.state = ACTIVE;
+                vector::push_back(&mut active_task_ids, task.task_index);
```
I understand that collecting active task ids, and then traversing them only for further calculation prevents us from some mistakes. but I think this mistakes can be/should be covered by unit-tests. So my suggestion would be not to create extra vector of `active-task-ids` here but to rely on the fact that the rest of the calculations are done after clean-up and any task that is traversed later is valid active task.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -237,35 +273,174 @@ module supra_framework::automation_registry {
 
-        // Apply the latest configuration if any parameter has been updated.
-        update_config_from_buffer();
+        let list = calculate_epoch_fees(
```
`calculate_tasks_automation_fees`

aregng replied:
`list -> tasks_automation_fees`

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -237,35 +273,174 @@ module supra_framework::automation_registry {
+            tcmg
+        );
+
+        let gas_committed_for_next_epoch = process_automation_task_fees(
```
`try_withdraw_task_automation_fees`

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -237,35 +273,174 @@ module supra_framework::automation_registry {
+        let min_interval = (math64::min(remaining_time, epoch_interval) as u256);
+
+        // Compute the base automation fee (taf). Total base fee for the interval
+        let taf = abf * task_max_gas * min_interval / max_gas_cap;
```
`task_occupancy_ratio_by_duration = min_interval * task_max_gas / max_gas_cap` can be calculated once and used in `taf and tcf` formulas

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -237,35 +273,174 @@ module supra_framework::automation_registry {
+
+        // Calculate congestion threshold surplus for the current epoch
+        let threshold_usage = (tcmg * DECIMAL / max_gas_cap) * 100;
+        let threshold_surplus = if (threshold_usage < threshold_percentage) 0 else threshold_usage - threshold_percentage;
```
maybe to avoid any extra calculations(multiplications) later we can return `0` directly here if the `threshold_usage < threshold_percentage`

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -237,35 +273,174 @@ module supra_framework::automation_registry {
+    /// Processes automation task fees by checking user balances.
+    /// - If the user has sufficient balance, deducts the fee and emits a success event.
+    /// - If the balance is insufficient, removes the task and emits a cancellation event.
+    fun process_automation_task_fees(
```
please note here that the tasks should be processed in the order of their registration. as removal actually does `swap-remove` then the order of the tasks will be mixed, so at some point `AutomationTaskFee` vector should be ordered based on the task id before doing charges. Let's also make sure that we have unit test covering the scenario where the tasks order is mixed.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -404,13 +567,15 @@ module supra_framework::automation_registry {
 
-        charge_automation_fee_from_user(
+        // Charge flate registration fee from the user at the time of registration
+        supra_account::transfer(
```
Just a reminder that we need to emit event related to `RegistrationFeeWithdraw`al as well. So charge and event can be in scope of separate function.

nizam-supraoracles committed:
remove `refund_automation_task_fee` function as we are not refunding

nizam-supraoracles committed:
create `cleanup_and_activate_tasks` function

nizam-supraoracles committed:
remove active_task_ids, and get active task from latest list to avoid mistakes

nizam-supraoracles committed:
rename few variables and function + optimize calculation

nizam-supraoracles committed:
update automation registry events and names, also include missing fee charges event

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -150,8 +156,16 @@ module supra_framework::automation_registry {
     #[event]
-    /// Withdraw user's registration fee event
-    struct FeeWithdrawnAdmin has drop, store {
+    /// Event on task registration fee withdrawal from owner account upon registration.
```
Line 154: canclled -> cancelled.
Line 152: Registration timestamp in seconds

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -150,8 +156,16 @@ module supra_framework::automation_registry {
+
+    #[event]
+    /// Emitted on withdrawal of specified amount from automation registry fee address to the sepcified address.
+    struct RegistryFeeWithdraw has drop, store {
```
sepcified -> specified

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -460,30 +657,7 @@ module supra_framework::automation_registry {
-            let refund_amount = residual_ttl * automation_registry_config.automation_base_fee_in_quants_per_sec;
-            transfer_fee_to_account_internal(user, refund_amount);
-            event::emit(
-                AutomationCancellationRefund { user, task_index: automation_task_metadata.id, amount: refund_amount }
```
`AutomationCancellationRefund` event can be removed as well

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -165,9 +179,55 @@ module supra_framework::automation_registry {
+    #[event]
+    /// Event emitted when an automation fee is refunded for an automation task at the end of the epoch for excessive
+    /// duration paid at the beginning of the epoch due to epoch-duration reduction by governance.
+    struct AutomationTaskFeeRefund has drop, store {
```
`Automation` prefix can be omitted.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -165,9 +179,55 @@ module supra_framework::automation_registry {
+    #[event]
+    /// Event emitted on automation task cancellation by owner.
+    struct TaskCancelled has drop, store {
+        task_index: u64
```
do you think if we add owner address here as well it will be redundant ?

nizam-supraoracles replied:
I guess, it's good to add owner address here as well

nizam-supraoracles committed:
remove duplicate fields in `AutomationTaskFee` struct

nizam-supraoracles committed:
add sorting before start fee withdraw from automation task

aregng reviewed (commented):
Just couple of cosmetic comments. I will do thorough review tomorrow :)

nizam-supraoracles committed:
automation registry minor nitpick

nizam-supraoracles committed:
autogenerated files update

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +280,186 @@ module supra_framework::automation_registry {
+        let acf = calculate_automation_congestion_fee(arc, tcmg);
+        let task_with_fees = vector[];
+
+        // Process each active task and apply the fee charges for the new epoch
```
comment is misleading, `and calculate fee for the epoch for the tasks`

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +280,186 @@ module supra_framework::automation_registry {
-        automation_registry.gas_committed_for_next_epoch = gas_committed_for_next_epoch;
-        automation_epoch_info.start_time = current_time;
-        automation_epoch_info.expected_epoch_duration = automation_epoch_info.epoch_interval;
+    /// Charges automation task fees for a single task at the time of new epoch.
```
Charges -> Calculates

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +280,186 @@ module supra_framework::automation_registry {
-        automation_epoch_info.expected_epoch_duration = automation_epoch_info.epoch_interval;
+    /// Charges automation task fees for a single task at the time of new epoch.
+    /// This is supposed to be called only after removing expired task and must not be called for expired task.
+    /// It's return
```
It's return ....

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +280,186 @@ module supra_framework::automation_registry {
+
+        let epoch_interval = aei.epoch_interval;
+
+        // Subtraction is safe here, as we already removed expiry tasks
```
expired tasks

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +280,186 @@ module supra_framework::automation_registry {
+        let threshold_usage = (tcmg * DECIMAL / max_gas_cap) * 100;
+        if (threshold_usage < threshold_percentage) 0
+        else {
+            let threshold_surplus = threshold_usage - threshold_percentage;
```
let's have `threshold_surplus` already normalized here i.e. `threshold_surplus_normalized = (threshold_usage - threshold_percentage) / 100
Othrewise the below statement looks to complicated.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +280,186 @@ module supra_framework::automation_registry {
+        let threshold_percentage = (arc.congestion_threshold_percentage as u256) * DECIMAL;
+
+        // Calculate congestion threshold surplus for the current epoch
+        let threshold_usage = (tcmg * DECIMAL / max_gas_cap) * 100;
```
let's have the following functions to do upscaling and downscaling to make code more readbale:
```
 fun upscale(value: u64) -> u256 {
    (value as u256) * DECIMAL
  }
  
  fun downscale(value: u256) -> u64 {
     (value / DECIMAL) as u64
 }
 
 ...
 let threshold_percentage = upscale(arc.congestion_threshold_percentage)
 ...
 let acf = arc.congestion_base_fee_in_quants_per_sec as u256) * threshold_surplus_norm;
 downscale(acf) 
 ```

nizam-supraoracles replied:
at the time of upscale, there are few parameters required type `u8` (arc.congestion_threshold_percentage) & `u256` (tcmg). do we need to create separate function for that `upscale_u256` & `upscale_u8`?

Same scenario for `downscale` as some return type required in `u256`, like `acf`

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +280,186 @@ module supra_framework::automation_registry {
+    ): u64 {
+        let gas_committed_for_next_epoch = 0;
+
+        vector_sorting(&mut tasks_automation_fees);
```
maybe `sort_by_task_index`

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -494,6 +658,24 @@ module supra_framework::automation_registry {
+                if (vector::borrow(v, j).task_index > vector::borrow(v, j + 1).task_index) {
+                    // Swap elements
+                    vector::swap(v, j, j + 1);
+                    vector::swap(v, j + 1, j);
```
This line seems redundant. Also let's try to cut it short, if no swapping is happening then the vector is already sorted.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +280,186 @@ module supra_framework::automation_registry {
+        // Subtraction is safe here, as we already removed expiry tasks
+        let remaining_time = task.expiry_time - current_time;
+        let min_interval = (math64::min(remaining_time, epoch_interval) as u256);
+        let task_occupancy_ratio_by_duration = min_interval * task_max_gas / max_gas_cap;
```
As discussed, any division that has numerator less than denominator  should have numerator upscaled and then downscaled.

nizam-supraoracles committed:
fix cosmetic comments

nizam-supraoracles committed:
implement automation epoch change fee refund

nizam-supraoracles committed:
registry module - add upscale and downscale methods

nizam-supraoracles committed:
automation registry - add `get_automation_epoch_info` method to fetch automation epoch info

nizam-supraoracles committed:
automation registry - add new filed to handle `epoch_freeze_fees` which is not withdrawable and provied view function for the same

nizam-supraoracles committed:
Merge branch 'feature/automation-networks' into task/issue-1421

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -206,6 +260,7 @@ module supra_framework::automation_registry {
             tasks: enumerable_map::new_map(),
             current_index: 0,
             gas_committed_for_next_epoch: 0,
+            epoch_freeze_fees: 0,
```
freeze -> locked

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
         let automation_epoch_info = borrow_global_mut<AutomationEpochInfo>(@supra_framework);
 
+        // Apply the latest configuration if any parameter has been updated.
+        update_config_from_buffer();
```
Config update should happen only after refund. The expected flow:
```
refund_tasks_for_previous_epoch(...)
update_config_from_buffer(...)
cleanup_and_activate_tasks(...)
charge_fees_for_new_epoch(...)
```

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
+        let max_gas_cap = (arc.registry_max_gas_cap as u256);
+
+        // Subtraction is safe here, as we already removed expired tasks
+        let remaining_time = task.expiry_time - current_time;
```
Here we need to be careful as this function is used to calculate refunds as well, so the task can be already expired (`task.expiry_time < current_time`) when refunds are calculated. In that case refund fee will be 0.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -494,13 +743,46 @@ module supra_framework::automation_registry {
 
+    #[view]
+    /// Get freeze balance of the resouce account
+    public fun get_epoch_freeze_balance(): u64 acquires AutomationRegistry {
```
resouce -> resource
freeze -> locked

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
+        current_time: u64
+    ) {
+        let epoch_duration = current_time - aei.start_time;
+        let residual_time = if (epoch_duration < aei.expected_epoch_duration) {
```
to avoid 2 `if` statements we can simply return here if `expected_epoch_duration` < epoch_duration.

nizam-supraoracles replied:
I think it should be like this `expected_epoch_duration <= epoch_duration`
```
if (aei.expected_epoch_duration <= epoch_duration) {
    return;
} else {
    let residual_time = aei.expected_epoch_duration - epoch_duration;
    ...
```

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
+    /// Calculates the total committed maximum gas for tasks that are not in the pending state.
+    fun calculate_total_committed_max_gas(automation_registry: &AutomationRegistry): u256 {
+        let total_committed_max_gas = 0;
+        let ids = enumerable_map::get_map_list(&automation_registry.tasks);
```
can we use here `enumerable_map::for_each_value_ref`

nizam-supraoracles replied:
Yes, perfect. We should use `enumerable_map::for_each_value_ref`, since enumerable_map provides this method

aregng reviewed (commented):
Will do one more review tomorrow.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
+        } else 0;
+
+        if (residual_time != 0) {
+            let tcmg = calculate_total_committed_max_gas(automation_registry);
```
`tcmg` calculated here and `tcmg` calculated during charge can differ if there were tasks that were cancelled due to insufficient balance or `automation-fee-cap` surpass.
I think we should keep/cache `tcmg` calculated during charge along with `epoch-locked-fee` to have refunds correctly calculated.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
+    }
+
+    /// Processes refunds for automation task fees.
+    fun process_task_fee_refunds(
```
may be `refund_tasks_fee`

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -143,31 +155,73 @@ module supra_framework::automation_registry {
+    /// Represents the fee charged for an automation task execution and some additional information.
+    struct AutomationTaskFee has drop {
+        task_index: u64,
+        owner: address,
```
I understand that you brought the owner back in order to not traverse registry tasks one more time. But I wonder whether the copy or traversal will be chipper in this case. Do we have means to check it ?

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
 
-        // Apply the latest configuration if any parameter has been updated.
-        update_config_from_buffer();
+    /// Charges automation task fees for all active tasks at the beginning of a new epoch.
```
/// Calculates automation task fees for the active tasks for the provided interval with provided `tcmg` occupancy. The `CANCELLED` tasks are also taken into account if `include_cancelled_task` is true.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
+    }
+
+    /// Calculates the total committed maximum gas for tasks that are not in the pending state.
+    fun calculate_total_committed_max_gas(automation_registry: &AutomationRegistry): u256 {
```
this function will be handy when we implement simulation view function to calculate automation-fee for the task with provided `max-gas-amount`.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
+        let task_occupancy_ratio_by_duration = (min_interval * upscale_from_u64(task.max_gas_amount)) / max_gas_cap;
+
+        // Compute the base automation fee (taf). Total base fee for the interval
+        let taf = downscale_to_u64(abf * task_occupancy_ratio_by_duration);
```
to do downscaling only once I would suggest to apply it on the `( taf+ tcf )`

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
+        else {
+            let threshold_surplus_normalized = (threshold_usage - threshold_percentage) / 100;
+            // Compute the automation congestion fee (acf) for the epoch
+            let acf = downscale_to_u256(
```
to make this statement more readable I would suggest to apply downscale on calculated acf 
``` downscale_to_u256(acf)```

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
+        }
+    }
+
+    /// Processes automation task fees by checking user balances.
```
and task's commitment on automation-fee, i.e. `automation-fee-cap`
- If calculated fee for the epoch surpasses task's `automation-fee-cap` task is removed and cancellation event is emitted.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -233,37 +288,257 @@ module supra_framework::automation_registry {
+        vector::for_each(tasks_automation_fees, |task| {
+            let task: AutomationTaskFee = task;
+            let task_metadata = enumerable_map::get_value(&automation_registry.tasks, task.task_index);
+            let user_balance = balance<SupraCoin>(task_metadata.owner);
```
I think here we can optimize the code, by first checking the `automation-fee-cap` if it is not surpassed only after that we can get and check against the balance of the owner

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1060,4 +1348,62 @@ module supra_framework::automation_registry {
+        assert!(balance == 9499928000, 12);
+    }
+
+    #[test(framework = @supra_framework, user = @0x1cafe)]
```
we need to have test-cases covering :
- task cancellation during automation-fee-charge
- task refund when epoch is shorter
- no task refund or extra fee charge when epoch is larger
- task refund when epoch is shorter and task is not yet expired
- task refund when epoch is shorter and task is has been already expired
- registry -fee withdraw with locked and not locked amount.

nizam-supraoracles committed:
Merge branch 'feature/automation-networks' of github.com:Entropy-Foundation/aptos-core into task/issue-1421

nizam-supraoracles committed:
rename `epoch_freeze_fees` to `epoch_locked_fees`

nizam-supraoracles committed:
store `total_committed_max_gas_amount` in the storage so the time of refund we will take it from here, + some consometic changes

nizam-supraoracles committed:
Merge branch 'task/issue-1421' of github.com:Entropy-Foundation/aptos-core into task/issue-1421

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -50,6 +54,10 @@ module supra_framework::automation_registry {
     const EDISABLED_AUTOMATION_FEATURE: u64 = 15;
+    /// Insufficient balance in the resource wallet for withdrawal
+    const EINSUFFICIENT_BALANCE: u64 = 16;
+    /// Requested amount exceeds the frozen balance
```
frozen -> locked

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -100,10 +110,14 @@ module supra_framework::automation_registry {
         gas_committed_for_next_epoch: u64,
+        /// Total fee charged to users during the epoch, which is not withdrawable
+        epoch_locked_fees: u64,
+        /// Total committed max gas amount for the epoch
```
at the beginning of the current epoch.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -100,10 +110,14 @@ module supra_framework::automation_registry {
+        /// Total fee charged to users during the epoch, which is not withdrawable
+        epoch_locked_fees: u64,
+        /// Total committed max gas amount for the epoch
+        total_committed_max_gas_amount: u256,
```
`gas_committed_for_this_epoch` or
`gas_committed_at_epoch_start` or
`occupancy_at_epoch_start`

nizam-supraoracles replied:
I am proceeding with `gas_committed_for_this_epoch`.

aregng commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -253,37 +311,259 @@ module supra_framework::automation_registry {
+        let epoch_duration = current_time - aei.start_time;
+        if (aei.expected_epoch_duration <= epoch_duration) {
+            return
+        } else {
```
I think we can omit else clause if we return in then clause.

nizam-supraoracles committed:
pr comments address

Commit:
Added unit tests and applied cosmentic changes

- added unit tests for automation fee charge and refund functionality

- updated fee-charge and refund functionality to take into account
  feature flag value and locked fee amount
    - Now if there is no locked fee amount no refund is attempted
    - If the feature is disabled refund. cleanup and config updates
      are done on new epoch but no task is charged automation fee for epoch as
      they will not be even scheduled for execution

- updated config update api to do check against epoch-interval and a new task
  duration cap. Now update will fail if epoch interval is greater than
  the new task-durationr- cap

- applied some cosmetic changes (formating, name-correction)

Commit:
Added view functions to estimate automation fees

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -197,6 +247,137 @@ module supra_framework::automation_registry {
+    #[view]
+    /// Extimates automation fee for the next epoch for specified task occupancey for the cofigured epoch-interval
+    /// referencing the current automation regitry fee parameters, current total occupancey and registry maximum allowed
+    /// occupancey for the next epoch.
```
- "Extimates" → "Estimates"
- "occupancey" → "occupancy" - (in 3 placed)
- "cofigured" → Should be "configured"
- "automation regitry" → "automation registry"

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -197,6 +247,137 @@ module supra_framework::automation_registry {
+    #[view]
+    /// Extimates automation fee the next epoch for specified task occupancey for the cofigured epoch-interval
+    /// referencing the current automation regitry fee parameters, specified total/committed occupancey and registry
+    /// maximum allowed occupancey for the next epoch.
```
Same typos, as mentioned above

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -261,115 +444,282 @@ module supra_framework::automation_registry {
+        });
+    }
+
+    /// Cleanup and actiavete the automation task also it's calculate and return total committed max gas
```
this typo is from my end
"actiavete" →"activate"

aregng replied:
It seems I was in rush to many types :1st_place_medal:

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -731,24 +1009,98 @@ module supra_framework::automation_registry {
     }
 
+    #[test_only]
+    /// Registers a task with specifed state and returns the task index
```
"specifed" →"specified"

nizam-supraoracles reviewed (commented):
Just a few typos identified. I will revisit the newly added unit test cases.

Commit:
Addressed review comments and updated unit tests

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1134,4 +1583,567 @@ module supra_framework::automation_registry {
+    ) acquires AutomationRegistry  {
+        initialize_registry_test(framework, user);
+        set_locked_fee(framework, 100_000_000);
+        let withdraw_amout = REGISTRY_DEFAULT_BALANCE + 1;
```
If we are testing the insufficient balance scenario, we don’t need `set_locked_fee` in this case.
also correct this everywhere, `amout` -> `amount`

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1134,4 +1583,567 @@ module supra_framework::automation_registry {
+
+    #[test(framework = @supra_framework, user = @0x1cafa)]
+    #[expected_failure(abort_code = EREQUEST_EXCEEDS_LOCKED_BALANCE, location = Self)]
+    fun check_registry_fee_failed_withdrawal(
```
Can we rename the function to something like `check_registry_fee_failed_withdrawal_locked_balance`, since the failure occurs due to locked balance?

Commit:
Fixed test failures

Commit:
Addressed review comments
